pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -272,11 +273,10 @@ func (s *Scheduler) add(ctx context.Context, pod *corev1.Pod) error {
 		}
 	}
 
-	// Consider using https://pkg.go.dev/container/heap
-	sort.Slice(s.newNodeClaims, func(a, b int) bool { return len(s.newNodeClaims[a].Pods) < len(s.newNodeClaims[b].Pods) })
-
-	// Pick existing node that we are about to create
-	for _, nodeClaim := range s.newNodeClaims {
+	h := NewNodeClaimHeap(s.newNodeClaims)
+	heap.Init(h)
+	for h.Len() > 0 {
+		nodeClaim := heap.Pop(h).(*NodeClaim)
```

Comment:
We should write some benchmarking tests to validate this makes things faster and by how much.

Reply:
+1

Reply:
I add some benchmark for evaluation, basically heap is better but not that much.
```sh
➜  karpenter git:(improve/nodeclaim-sorting) ✗ go test -run=none -bench=Sorting -benchmem sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling -v
goos: linux
goarch: amd64
pkg: sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling
cpu: Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz
BenchmarkHeapSorting
BenchmarkHeapSorting-4          19477929                86.49 ns/op           24 B/op          1 allocs/op
BenchmarkSliceSorting
BenchmarkSliceSorting-4          8984490               116.3 ns/op            24 B/op          1 allocs/op
PASS
ok      sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling   4.877s
➜  karpenter git:(improve/nodeclaim-sorting) ✗ 
➜  karpenter git:(improve/nodeclaim-sorting) ✗ go test -run=none -bench=Sorting -benchmem sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling -v
goos: linux
goarch: amd64
pkg: sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling
cpu: Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz
BenchmarkHeapSorting
BenchmarkHeapSorting-4          16447299                62.18 ns/op           24 B/op          1 allocs/op
BenchmarkSliceSorting
BenchmarkSliceSorting-4         13348653                78.73 ns/op           24 B/op          1 allocs/op
PASS
ok      sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling   2.304s
➜  karpenter git:(improve/nodeclaim-sorting) ✗ go test -run=none -bench=Sorting -benchmem sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling -v
goos: linux
goarch: amd64
pkg: sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling
cpu: Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz
BenchmarkHeapSorting
BenchmarkHeapSorting-4          17975103                58.27 ns/op           24 B/op          1 allocs/op
BenchmarkSliceSorting
BenchmarkSliceSorting-4         13778595                88.28 ns/op           24 B/op          1 allocs/op
PASS
ok      sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling   2.478s
➜  karpenter git:(improve/nodeclaim-sorting) ✗ go test -run=none -bench=Sorting -benchmem sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling -v
goos: linux
goarch: amd64
pkg: sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling
cpu: Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz
BenchmarkHeapSorting
BenchmarkHeapSorting-4          20438842                59.15 ns/op           24 B/op          1 allocs/op
BenchmarkSliceSorting
BenchmarkSliceSorting-4         15978789                75.94 ns/op           24 B/op          1 allocs/op
PASS
ok      sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling   6.341s
➜  karpenter git:(improve/nodeclaim-sorting) ✗ go test -run=none -bench=Sorting -benchmem sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling -v
goos: linux
goarch: amd64
pkg: sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling
cpu: Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz
BenchmarkHeapSorting
BenchmarkHeapSorting-4          17908522                71.79 ns/op           24 B/op          1 allocs/op
BenchmarkSliceSorting
BenchmarkSliceSorting-4         13740463                77.62 ns/op           24 B/op          1 allocs/op
PASS
ok      sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling   2.566s
➜  karpenter git:(improve/nodeclaim-sorting) ✗ go test -run=none -bench=Sorting -benchmem sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling -v
goos: linux
goarch: amd64
pkg: sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling
cpu: Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz
BenchmarkHeapSorting
BenchmarkHeapSorting-4          16799110                67.39 ns/op           24 B/op          1 allocs/op
BenchmarkSliceSorting
BenchmarkSliceSorting-4         13493124                97.02 ns/op           24 B/op          1 allocs/op
PASS
ok      sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling   4.522s
➜  karpenter git:(improve/nodeclaim-sorting) ✗ 
```